compiler/src/dotty/tools/dotc/transform/TupleOptimizations.scala

```diff
@@ -25,17 +25,17 @@ class TupleOptimizations extends MiniPhase with IdentityDenotTransformer {
 
   override def transformApply(tree: tpd.Apply)(implicit ctx: Context): tpd.Tree =
     if (!tree.symbol.exists || tree.symbol.owner != defn.DynamicTupleModuleClass) tree
-    else if (tree.symbol == defn.DynamicTuple_dynamicCons) transformTupleCons(tree)
-    else if (tree.symbol == defn.DynamicTuple_dynamicTail) transformTupleTail(tree)
-    else if (tree.symbol == defn.DynamicTuple_dynamicSize) transformTupleSize(tree)
-    else if (tree.symbol == defn.DynamicTuple_dynamicConcat) transformTupleConcat(tree)
-    else if (tree.symbol == defn.DynamicTuple_dynamicApply) transformTupleApply(tree)
-    else if (tree.symbol == defn.DynamicTuple_dynamicToArray) transformTupleToArray(tree)
+    else if (tree.symbol == defn.DynamicTuple_cons) transformTupleCons(tree)
+    else if (tree.symbol == defn.DynamicTuple_tail) transformTupleTail(tree)
+    else if (tree.symbol == defn.DynamicTuple_size) transformTupleSize(tree)
+    else if (tree.symbol == defn.DynamicTuple_concat) transformTupleConcat(tree)
+    else if (tree.symbol == defn.DynamicTuple_apply) transformTupleApply(tree)
+    else if (tree.symbol == defn.DynamicTuple_toArray) transformTupleToArray(tree)
     else tree
 
   private def transformTupleCons(tree: tpd.Apply)(implicit ctx: Context): Tree = {
     val head :: tail :: Nil = tree.args
-    defn.tupleTypes(tree.tpe) match {
+    defn.tupleTypes(tree.tpe.widenTermRefExpr.dealias) match {
```

Comment:
Maybe move `.dealias` before `. widenTermRefExpr`? Can we just use `widen` instead of `widenTermRefExpr`?

Reply:
No, the alias appears after the widening of the term ref. I did not find a case where we needed to completely widen the type. For the case of indices and sizes, this might widen the constant typea.

Reply:
We only widen the term refs that come from inserting val bindings when inlining.

Reply:
We could mark the parameters an inline to avoid it. But I will remove the inline altogether in a future PR as it does not really have a purpose anymore.

Reply:
> For the case of indices and sizes, this might widen the constant type.

The type is a tuple type, it thus may not be constant types.

For dealiasing, what I've in mind is `type T = a.type`. But I guess no programmers will write such code, so it's fine to ignore such cases.